openassessment/data.py

```diff
@@ -670,3 +676,178 @@ def collect_ora2_responses(cls, course_id, desired_statuses=None):
+    to submissions (attachments, answer texts).
+    """
+
+    ATTACHMENT = 'attachment'
+    TEXT = 'text'
+
+    @classmethod
+    def _download_file_by_key(cls, key):
+        download_url = urljoin(
+            settings.LMS_ROOT_URL, get_download_url(key)
+        )
+
+        response = requests.get(download_url)
+        response.raise_for_status()
+
+        return response.content
+
+    @classmethod
+    def _create_downloads_csv(cls, submission_files_data):
+        """
+        Creates csv where each row represents all information about some certain
+        submission attachment or text.
+        """
+
+        header = (
+            'Course Id',
+            'Block Id',
+            'Student Id',
+            'File Key',
+            'File Name',
+            'File Type',
+            'File Description',
+            'File Size',
+            'Path',
+        )
+
+        rows = []
+
+        for file_data in submission_files_data:
+            rows.append(
+                (
+                    file_data['course_id'],
+                    file_data['block_id'],
+                    file_data['student_id'],
+                    file_data['key'],
+                    file_data['name'],
+                    file_data['type'],
+                    file_data['description'],
+                    file_data['size'],
+                    file_data['file_path'],
+                )
+            )
+
+        output_buffer = StringIO()
+        csvwriter = csv.writer(output_buffer)
+        csvwriter.writerows([header] + rows)
+
+        return output_buffer
+
+    @classmethod
+    def create_zip_with_attachments(cls, file, course_id, submission_files_data):
+        """
+        Opens given stream as a zip file and writes into it all submission
+        attachments and csv with list of all downloads.
+
+        Example of result zip file structure:
+        ```
+        .
+        └── CourseId
+            ├── BlockId1
+            │   ├── StudentId1
+            │   │   ├── attachments
+            │   │   │   ├── SomeFile1
+            │   │   │   └── SomeFile2
+            │   │   ├── part_0.txt
+            │   │   └── part_1.txt
+            │   └── StudentId2
+            │       ├── part_0.txt
+            │       └── part_1.txt
+            ├── BlockId2
+            │   └── StudentId3
+            │       ├── attachments
+            │       │   └── SomeFile4
+            │       └── part_0.txt
+            └── downloads.csv
+        ```
+        """
+        with ZipFile(file, 'w') as zip_file:
+            for file_data in submission_files_data:
+                file_content = (
+                    cls._download_file_by_key(file_data['key'])
+                    if file_data['type'] == cls.ATTACHMENT
+                    else file_data['content']
+                )
+
+                zip_file.writestr(file_data['file_path'], file_content)
+
+            downloads_csv = cls._create_downloads_csv(submission_files_data)
```

Comment:
If the download csv is only ever used in this zip file, why not just create the csv as you iterate `submission_files_data`?